File: src/main/java/com/official/memento/alarm/infrastructure/client/AlarmSendClientAdapter.java
package com.official.memento.alarm.infrastructure.client;

import com.official.memento.alarm.domain.port.AlarmOutputPort;
import com.official.memento.global.stereotype.Adapter;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.MediaType;
import org.springframework.scheduling.annotation.Async;
import org.springframework.web.reactive.function.client.WebClient;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.List;

@Adapter
@RequiredArgsConstructor
public class AlarmSendClientAdapter implements AlarmOutputPort {

    private final WebClient webClient;

    @Value("${discord.error-alert-url}")
    private String discordErrorAlarmUri;

    @Override
    @Async
    public void sendAlarm(String uri, String content) {
        webClient.post()
                .uri(uri)
                .header("Content-Type", MediaType.APPLICATION_JSON_VALUE)
                .bodyValue(new AlarmRequestBody(content, null))
                .retrieve()
                .toBodilessEntity()
                .subscribe();
    }

    @Override
    @Async
    public void sendExceptionAlarm(Exception e) {
        webClient.post()
                .uri(discordErrorAlarmUri)
                .header("Content-Type", MediaType.APPLICATION_JSON_VALUE)
                .bodyValue(new AlarmRequestBody(
                        "에러 발생 : " + e.getMessage() + " ( " + e.getClass().getSimpleName() + " )",
                        List.of(new Embed(
                                "Stack Trace",
                                parseStackTrace(e)
                        ))
                ))
                .retrieve()
                .toBodilessEntity()
                .subscribe();
    }

    private String parseStackTrace(Exception exception) {
        StringWriter sw = new StringWriter();
        PrintWriter pw = new PrintWriter(sw);
        exception.printStackTrace(pw);
        String stackTrace = sw.toString();

        if (stackTrace.length() > 2000) {
            return stackTrace.substring(0, 2000);
        }
        return stackTrace;
    }

    record AlarmRequestBody(
            String content,
            List<Embed> embeds
    ) {}

    record Embed(
            String title,
            String description
    ) {}
}

File: src/main/kotlin/com/official/memento/alarm/infrastructure/client/AlarmSendClientAdapter.kt
package com.official.memento.alarm.infrastructure.client

import com.official.memento.alarm.domain.port.AlarmOutputPort
import com.official.memento.global.stereotype.Adapter
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.MediaType
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBodilessEntity

@Adapter
class AlarmSendClientAdapter(
    private val webClient: WebClient,
    @Value("\${discord.error-alert-url}")
    private val discordErrorAlarmUri: String
) : AlarmOutputPort {
    override fun sendAlarm(
        uri: String,
        content: String,
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            webClient.post()
                .uri(uri)
                .header("Content-Type", MediaType.APPLICATION_JSON_VALUE)
                .bodyValue(AlarmRequestBody(content= content, embeds = null))
                .retrieve()
                .awaitBodilessEntity()
        }
    }

    override fun sendExceptionAlarm(e: Exception) {
        CoroutineScope(Dispatchers.IO).launch {
            webClient.post()
                .uri(discordErrorAlarmUri)
                .header("Content-Type", MediaType.APPLICATION_JSON_VALUE)
                .bodyValue(
                    AlarmRequestBody(
                    content = "에러 발생 : ${e.message} ( ${e.javaClass.simpleName} )",
                    listOf(
                        Embed(
                            title = "Stack Trace",
                            description = parseStackTrace(e),
                        ),
                    ),
                ))
                .retrieve()
                .awaitBodilessEntity()
        }
    }

    data class AlarmRequestBody(
        val content: String,
        val embeds: List<Embed>?
    )


    data class Embed(
        val title: String,
        val description: String,
    )

    private fun parseStackTrace(exception: Exception): String {
        val stackTrace = exception.stackTraceToString()
        if (stackTrace.length > 2000) {
            return stackTrace.substring(0, 2000)
        }
        return stackTrace
    }
}
